src/data/ast.ts

```diff
@@ -9,69 +10,134 @@ export default function sqlToAST(sql: string): Map<string, Clause> {
+    // add the phrase to the bracket phrase. The complete bracket phrase will be used to create a new AST branch
     if (bracketCount > 0) {
       bracketPhrase += foundNode + phrase;
     } else {
-      ast.set(foundNode, phrase);
+      ast.set(foundNode, [phrase]);
       lastNode = foundNode;
+      bracketPhrase = phrase;
     }
     bracketCount += (phrase.match(/\(/g) || []).length;
     bracketCount -= (phrase.match(/\)/g) || []).length;
     if (bracketCount <= 0 && lastBracketCount > 0) {
-      // The phrase brackets is complete
-      // If it is a FROM phrase lets make a branch node
-      // If it is anything else lets make a leaf node
-      if (lastNode === 'FROM') {
-        ast.set(lastNode, sqlToAST(bracketPhrase));
+      // The phrase is complete
+      // If it contains the keyword SELECT, make new branches
+      // If it does not, make a leaf node
+      if (bracketPhrase.match(/\bSELECT\b/gi)) {
+        ast.set(lastNode, getASTBranches(bracketPhrase));
       } else {
-        const p = (ast.get(lastNode) as string).concat(bracketPhrase);
-        ast.set(lastNode, p);
+        ast.set(lastNode, [bracketPhrase]);
       }
     }
     lastBracketCount = bracketCount;
   }
   return ast;
 }
 
-export function clausesToSql(ast: Map<string, Clause>): string {
+export function ASTToSql(ast: AST): string {
   let r = '';
-  ast.forEach((c: Clause, key: string) => {
-    if (typeof c === 'string') {
-      r += `${key} ${c.trim()} `;
-    } else if (c !== null) {
-      r += `${key} (${clausesToSql(c)} `;
+  ast.forEach((clauses: Clause[], key: string) => {
+    let keyAndClauses = `${key} `;
+    for (let c of clauses) {
+      if (typeof c === 'string') {
+        keyAndClauses += `${c.trim()} `;
+      } else if (c !== null) {
+        keyAndClauses += `${ASTToSql(c)} `;
+      }
+    }
+    // do not add the keys that do not have nodes
+    if (keyAndClauses !== `${key} `) {
+      r += keyAndClauses;
     }
   });
   // Remove all of the consecutive spaces to make things more readable when debugging
   return r.trim().replace(/\s+/g, ' ');
 }
 
-// Creates a statement with all the keywords to preserve the keyword order
-function createStatement() {
-  const clauses = new Map<string, Clause>();
-  clauses.set('WITH', null);
-  clauses.set('SELECT', null);
-  clauses.set('DISTINCT', null);
-  clauses.set('FROM', null);
-  clauses.set('SAMPLE', null);
-  clauses.set('JOIN', null);
-  clauses.set('PREWHERE', null);
-  clauses.set('WHERE', null);
-  clauses.set('GROUP BY', null);
-  clauses.set('LIMIT BY', null);
-  clauses.set('HAVING', null);
-  clauses.set('LIMIT', null);
-  clauses.set('OFFSET', null);
-  clauses.set('UNION', null);
-  clauses.set('INTERSECT', null);
-  clauses.set('EXCEPT', null);
-  clauses.set('INTO OUTFILE', null);
-  clauses.set('FORMAT', null);
+function getASTBranches(sql: string): Clause[] {
+  const clauses: Clause[] = [];
+  const re = /\b(AND|OR|,)\b/gi;
+  let bracketCount = 0;
```

Comment:
I like this! I think it would be best if it was `const bracket = {count, lastCount, phrase}` because regExp, index and lastPhraseIndex do not have a strong relation with bracket

Reply:
maybe the object is called ParserState or something

Reply:
I think it looks a lot better with the new `bracket` var. Do you think we still need a parserState object? I'm not strongly opinionated either way

Reply:
its not needed,  maybe a bit cleaner to have the state in one object